File: solutions/3876-construct-uniform-parity-array-ii/counting.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
  public:
    bool uniformArray(vector<int> &nums1) {
        int evenMin = 0;
        int oddMin = 0;

        for (const int num : nums1) {
            if (num & 1) {
                if (oddMin == 0)
                    oddMin = num;
                else
                    oddMin = min(oddMin, num);
            } else {
                if (evenMin == 0)
                    evenMin = num;
                else
                    evenMin = min(evenMin, num);
            }
        }

        if (!evenMin || !oddMin || oddMin < evenMin) return true;
        return false;
    }
};

File: solutions/3876-construct-uniform-parity-array-ii/counting.js
// time: O(n)
// space: O(1)

/**
 * @param {number[]} nums1
 * @return {boolean}
 */
const uniformArray = nums1 => {
    let evenMin = Infinity;
    let oddMin = Infinity;

    for (const num of nums1) {
        if (num & 1) oddMin = Math.min(oddMin, num);
        else evenMin = Math.min(evenMin, num);
    }

    if (oddMin === Infinity || evenMin === Infinity || oddMin < evenMin)
        return true;
    return false;
};

